### DISCUSSION TITLE
Feasibility of a TFM for Fable

### DISCUSSION THREAD

**Discussion by jwosty** [NuGet/Home#12965] [Ideas]
Hello NuGet team, we're currently exploring the possibility of a custom MSBuild SDK for [Fable](https://fable.io), the very popular F# transpiler (usually targeting Javascript, but more recently supports others). There are many Fable compatible packages already out there ([see here](https://fable.io/packages/)), and I think they would benefit greatly from having Fable-related TFMs. One such benefit would be a NuGet-blessed way to communicate Fable support rather than the adhoc system hijacking package tags currently used (see https://fable.io/docs/your-fable-project/author-a-fable-library.html). Another benefit would be a more unified package graph taking advantage of multitargeting, as most Fable packages end up distributing two parallel copies (one for .NET, one for Fable) for various reasons I won't get into here.

For example, take [the Elmish project](https://github.com/elmish/elmish). Today, it distributes two package: [Elmish](https://www.nuget.org/packages/Elmish), and [Fable.Elmish](https://www.nuget.org/packages/Fable.Elmish). They are logically the exact same thing, just that one is the .NET version and the other is the Fable version. If we had Fable TFMs, it could be distributed as one single `Elmish` package with two target frameworks: `netstandard2.0` and `fable4.0`.

Take another example, [Fable.React](https://github.com/fable-compiler/fable-react) (Fable bindings for the Javascript React library). It could have one TFM of `fable4.0-javascript`.

For context, please refer to our existing discussion at https://github.com/fable-compiler/Fable/issues/3549. Specifically, [this comment](https://github.com/fable-compiler/Fable/issues/3549#issuecomment-1775991166) explores some potential TFM strategies, and we would love feedback on which way fits better. There's also a proof-of-concept SDK here: https://github.com/jwosty/Fable.MSBuild.Sdk.

So, is this something NuGet would even consider? If so, what would you need from us to make this happen? Does a design need to be submitted somewhere? Perhaps some compatibility rules need to be spelled out?

I want to be clear - we're still in the research phase; this isn't us formally asking for this feature yet.

@MangelMaxime @baronfel @nojaf

**Comment by jwosty** [comment]
The SDK linked also experiments with broader MSBuild integration. For those who are not familiar with Fable, today you invoke it as a dotnet CLI tool (https://www.nuget.org/packages/Fable). It scrapes your `.fsproj` for sources files and dependencies and transpiles everything to Javascript (or another language).

The custom SDK will serve the purpose of integrating this process into `dotnet build`.

**Comment by MangelMaxime** [comment]
Hello,

Could someone from the NuGet team tell us if it is possible to create TFM for Fable like done for Tizen, Android, iOS, tvOS ?
